New Field Type: Address / Map

sc0ttkclark opened the pull request:
Patched from https://github.com/pods-framework/pods-address-maps
#657

Bringing over PR from #3623 
## Address field type
- [x] Name field type "address" (type) / "Address" (label)
- [x] Display in frontend with HTML markup
- [x] Custom HTML markup possible with magic tags ( `{{address_tag}}` )
- [x] Option to format in schema.org microdata
- [ ] Extend the province/region options. @JoryHogeveen (related: #4116)
  - [ ] Add an option to set this as a select box instead of a text input. (currently only supports US states)
  - [ ] Add a callback function that looks for a list file with existing provinces/region options by country (parameter) + add a filter to change/extend this list.
  - [ ] Add a small AJAX script that autofills the available options based on the above callback.

## Maps component
- [x] "Maps" component name/label
- [x] Hook into field type options to add Maps-specific options (PR #3711 is required)

**Backend UI** 
- [x] Base (currently loads the map with the correct address
- [x] Hidden lat/lng fields when `address` or `free-form` display is selected
- [x] Store lat/lng values after call to geocoder (even if these fields are not visible) to decrease number of api calls
- [x] Determine where API call should and shouldn't be done

**Frontend UI**
- [x] Display address field as a map on front end with use of the providers
- [x] I'd say it's important that this plugin returns an array instead of the display value when using `get_post_meta`, possible workaround for this?
- [x] Fix: Array storage triggers `display()` multiple times
- [ ] Option to set width/height @JoryHogeveen
- [ ] Option to use a Pods Template for the infowindows / marker-tooltip. @JoryHogeveen

**API**
- [ ] API methods for sorting objects based on GEO - [wp ticket](https://wordpress.org/support/topic/field-type-google-map-location/#post-10608913)
- [ ] Map display options for object relationships or a list of objects @JoryHogeveen
  - [x] Make base front-template work with multiple values (basics) @JoryHogeveen
  - [ ] Option to order by distance (straight line can be done without API) - [example](https://stackoverflow.com/questions/10053358/measuring-the-distance-between-two-coordinates-in-php)

### Providers
#### Google Maps API handler (priority)
- [x] Default geocode calls
- [ ] Check "Client ID" feature (API for business) https://developers.google.com/maps/premium/overview#client-id

**Display**
- [x] Load the map in backend
- [x] Show markers
- [x] Update fields when moving the marker
- [x] Update GEO data when changing the field values (freeform & address input type)
- [x] Show infowindows
- [x] Display address values or custom input data in infowindow
- [x] Custom markers (field type `file` not working in the field edit tab @sc0ttkclark )
- [ ] Map style (interactive map / static image)
- [x] Map type (terrain, satellite, etc)
- [x] Front-end display (based on back-end field display, but then with limited functions etc)
- [x] Option to disable scrolling @JoryHogeveen

# General
- [x] Determine storage structure @sc0ttkclark 
- [x] Determine validation @sc0ttkclark 
- [ ] Fix todo's @sc0ttkclark 
  - fields/address.php
    - [ ] L202 - Storage types
    - [ ] L286 - Validate each returned value for variable type and content (sanitizing)
    - [ ] L414 - check pregreplace, maybe this can be done better (nl2br not working)
  - Maps.php (component)
    - [ ] L87 (assets) - Allways load required front end assets: Maybe as an option?
    - [ ] L217 - Put part of the general maps config in the maps provider class? This could be different between providers.
    - [ ] L311 - Put maps field options in the maps provider class? This could be different between providers.
    - [ ] L329 - Autocorrect is not implemented yet (maybe old?)
    - [ ] L409 - 'display_type' is only available for field type 'address'
    - [ ] L535 (validation) - Validate based on address type ( lat / lon, address fields)
  - [ ] front/map-google.php + field/map-google.php - Check field type: support other field types for maps?
- [x] Delete old/unused files
- [ ] Tests!

# Future tasks

**Structure**
- [ ] MV-field structure (+ maps logic in a JS object) @pglewis 
- [ ] Global maps JS file instead of inline within the field/front files

**Other API integrations**
- [ ] Bing Maps API handler (needs research + contributor!)
- [ ] OpenStreetMap API handler (needs research + contributor!)
- [ ] Leaflet API handler (needs research + contributor!) http://leafletjs.com/index.html

JoryHogeveen committed:
First commit for address/map field patched from https://github.com/pods-framework/pods-address-maps

JoryHogeveen commented:
This field is not ready for full review yet, maybe remove that label.

I'll put the same question here for clarity:

@sc0ttkclark @pglewis 
Why not set the regular address fields as default (no map options etc) and add the map options with a component.
This way the address field is a regular field with multiple inputs and a component can extend this field with extra options for Google Maps or even order map providers.

The default address field would be quite similar to the GF address field this way (just regular fields and select boxes)
Let me know your thoughts on this.

sc0ttkclark commented:
Still needs review of PR and that question

JoryHogeveen committed:
First version of Address field

- Add address in backend
- Display address in frontend with HTML markup (config with format)
- Custom HTML markup format possible with magic tags
- Optionally add schema.org microdata

JoryHogeveen committed:
Improve microdata handler

JoryHogeveen committed:
First commit for Maps component

Currently only handles PodsAdmin fields
Requires PR: #3711

JoryHogeveen committed:
Improve HTML and microdata formatting

Also adds classes for use in front-end prepended with `pods-address`

JoryHogeveen committed:
Improve UX and options for Maps component

JoryHogeveen committed:
First commit for Google Maps API class

JoryHogeveen committed:
plugin_dir_path already has a trailing slash

JoryHogeveen committed:
Prepend option names with "map_"

JoryHogeveen committed:
var api_key needs to be static + api_key disable apikey for geocode testing

JoryHogeveen committed:
Add assets

JoryHogeveen committed:
Base of backend field working

- All maps additions are now in the component folder for consistency
- Main component handles default (global) calls and redirects them to
the right provider class
- Field UI: Use the localization JS object and extend this for more
usages like ajax calls to the main component for geodata
- Field UI: Geodata lookup when typing a new address
- Provider: API key is recently required for Google Maps

JoryHogeveen committed:
Just the old file, this is where the frond-end code will go

JoryHogeveen committed:
Generate hidden lat/lng fields for geo storage on the field UI

JoryHogeveen committed:
Patch the lat/lng field to the component UI directory + fix field types

JoryHogeveen committed:
JS fixes for geodata

- Bring back "lookup" button, seems like a better approach
- Validate the view type that this map field is appended to

JoryHogeveen committed:
geocoder url can be public + append json by default

JoryHogeveen committed:
Put lat/lng data in its own array key (geo)

JoryHogeveen committed:
Geo data validation (serverside) + main geocoder methods that call provider methods

JoryHogeveen committed:
Rebuild Maps field JS

- Add markers
- Add marker and map animations
- Updating address also updates map and marker
- Moving marker also updates address fields with new location

TODO: @sc0ttkclark @pglewis  Review JS code

JoryHogeveen committed:
pods-maps.js and its JS object (PodsMaps) are probably not needed anymore in field UI

JoryHogeveen committed:
Output options for pick type (regions & countries)

JoryHogeveen committed:
Put output options formatting in dedicated function

JoryHogeveen committed:
Docs, comments

JoryHogeveen committed:
dragend event instead of mouseup

JoryHogeveen committed:
Maps in full with + padding for labels

JoryHogeveen committed:
InfoWindow basics

JoryHogeveen committed:
@todo check pregreplace, maybe this can be done better (nl2br not working)

JoryHogeveen committed:
InfoWindow custom content option dependencies

JoryHogeveen committed:
Update description style when it's directly after a label

JoryHogeveen committed:
InfoWindow explanation for magictags

JoryHogeveen committed:
InfoWindows done

JoryHogeveen committed:
Comment sections

JoryHogeveen committed:
2 custom input options for infowindows (paragraph, wysiwyg) + set option values as field types

JoryHogeveen committed:
Add validate & pre_save filters

JoryHogeveen committed:
Remove unused code (this is now in the Maps component)

JoryHogeveen committed:
Freeform input type within address field view

JoryHogeveen committed:
Freeform input + JS improvements

JoryHogeveen commented:
**Note:**
Old files that are to be deleted since they are now (or will be) part of the Maps component
- classes / fields / addressmap.php
- ui / css / pods-address-maps.css
- ui / fields / lat-long.php
- ui / fields / map.php
- ui / front / address-map.php
- ui / front / address.php (is in fields)
- ui / front / lat-long.php
- ui / front / maps.php
- ui / js / pods-address-maps.js

JoryHogeveen commented on classes/fields/address.php:
```diff
@@ -0,0 +1,508 @@
+				'type'       => 'pick',
+				'data'       => array(
+					'long' => __( 'Full name', 'pods' ),
+					'short' => __( 'Stage / Province code', 'pods' )
```
Typo: Stage -> State

JoryHogeveen committed:
Make field label translatable

See #3849

JoryHogeveen committed:
Rename map.php to map-google.php

JoryHogeveen committed:
php linter

JoryHogeveen committed:
Defai;t type is address

JoryHogeveen committed:
Prevent overwriting the $value var

JoryHogeveen committed:
Implement frontend display hook basics

JoryHogeveen committed:
Add more todo's + prepare provider specific js files

JoryHogeveen commented:
Found an issue that needs to be addressed:

When a field returns a array from serialised stored, in this case it gets looped through in PodsForm::display() and causes the front-end logic to break the value.

@sc0ttkclark I see there is a $tableless_field_types check for this. Can you give me some info on what you exactly consider a "tableless field type"?

JoryHogeveen committed:
Merge remote-tracking branch 'refs/remotes/origin/release/2.7' into feature/#657_address_map_field

JoryHogeveen committed:
Remove field type from maps options

These are options for all compatible field types (future)

JoryHogeveen committed:
Basic front-end maps are now working

JoryHogeveen committed:
Remove dev leftovers

JoryHogeveen committed:
Improve naming and structure

+ remove unused stuff

sc0ttkclark committed:
Merge branch 'release/2.7' into feature/#657_address_map_field

sc0ttkclark committed:
Merge branch 'release/2.7' into feature/#657_address_map_field

sc0ttkclark committed:
Ensure add button is i18n'd

sc0ttkclark committed:
Hack around component setting saving for file field types, the post data was multidimensional array with 'id' keys

sc0ttkclark committed:
Fix phpdoc

sc0ttkclark committed:
Fix selected handling for single-value or multi-value simple pick fields

sc0ttkclark committed:
Add display_list handling (just display once, not for every key in our field value)

sc0ttkclark committed:
Add output escaping and get image url from marker file ID

sc0ttkclark committed:
Whitespace tweak

sc0ttkclark committed:
Handle file/pick array needs for saving single/multi values

sc0ttkclark committed:
Remove some unneeded files

sc0ttkclark committed:
Formatting stuff

sc0ttkclark committed:
Handle output display option

sc0ttkclark committed:
code formatting and output escaping for address

JoryHogeveen committed:
Fix geo lookup location button

JoryHogeveen committed:
Marker icon field input + improve var declaration

JoryHogeveen committed:
Marker icon display

JoryHogeveen committed:
Fix map type display

JoryHogeveen committed:
Set infowindow triggers on display field

JoryHogeveen commented:
@sc0ttkclark 
About the data storage. Currently it's all stored serialized under one metakey. Do you want to support multiple metakeys for the different data as well?

JoryHogeveen committed:
Add filters to add your own custom maps provider

JoryHogeveen committed:
Add todo's for maps options. These can be different for each provider.

JoryHogeveen committed:
Better method names for providers

JoryHogeveen committed:
Improve code readability

sc0ttkclark committed:
Merge branch 'release/2.7' into feature/#657_address_map_field

sc0ttkclark committed:
Merge branch 'release/2.7' into feature/#657_address_map_field

# Conflicts:
#	classes/PodsComponents.php
#	classes/fields/pick.php

sc0ttkclark commented:
@JoryHogeveen Just fixed the conflicts, can you come back around on the tasklist in this PR and update it with whatever's not specifically called out? If there are `@todo` markers that we need to take care of, can you add them as tasklist items here?

JoryHogeveen commented:
@sc0ttkclark 
Todolist stayed the same. I've added the todo's I saw in the code to the list.
I think the main topic for now is a structure discussion, both in PHP and JS it's important that the component is flexible to work with for other map providers.

sc0ttkclark committed:
Merge branch 'release/2.7' into feature/#657_address_map_field

JoryHogeveen committed:
Use pods_v()

JoryHogeveen committed:
Option to disable the scrollwheel + use pods_v()

JoryHogeveen committed:
Refactor google maps display to work with multiple values.

JoryHogeveen committed:
Typo

JoryHogeveen committed:
Allow values to overwrite the field marker

JoryHogeveen committed:
Ensure geodata is passed as float values

jimtrue commented:
@JoryHogeveen is this going to include the GeoMashup (Archive of all fields with the 'map' field or related items, etc. being displayed on an overview map). I saw you mention it in #3982 but didn't see it linked through

JoryHogeveen commented:
@jimtrue That is one of the todolist items yes.

JoryHogeveen committed:
Maps provider interface

JoryHogeveen committed:
Move provider related map options into the provider class

JoryHogeveen committed:
Get template titles helper

JoryHogeveen committed:
Autocomplete option + filter

JoryHogeveen committed:
Enhance error handling

JoryHogeveen committed:
autocorrect option setting

JoryHogeveen committed:
Fix method name

JoryHogeveen committed:
Check display setting, use pods_v & phpdoc

JoryHogeveen committed:
Only return string values for address parts

JoryHogeveen committed:
Google address data handler enhancements + filter

JoryHogeveen committed:
Try again if Google failes the first time

JoryHogeveen committed:
phpDoc

JoryHogeveen committed:
Option handler fixes

JoryHogeveen committed:
Code format

JoryHogeveen committed:
Handle multiple values

JoryHogeveen committed:
Option to use templates as infowindow content

JoryHogeveen committed:
Option to combine equal geo locations (!! no UI for this yet !!)

JoryHogeveen committed:
map container atts

JoryHogeveen committed:
Autoclose open infowindows if a new window is clicked

JoryHogeveen committed:
Autozoom when there are multiple values (!! no UI for this yet !!)

Jen340922 commented:
Has this been released yet?

tiptronic commented:
Is there any progress or info on this?

bwdgraphics commented:
Hi. I'm looking for help importing from a CSV into this PODs "address" field. I see the fields within the field for the data I want to import (Addr 1, Addr 2, City, State, Zip) but not sure how to map the import to put the data into those fields. Have tried a few ways, but no luck so far. How would I go about doing this?

jimtrue commented:
@bwdgraphics I responded over on the WordPress.org forums; what are you using to perform the import with?

bwdgraphics commented:
Yes, that's how I arrived here, actually. We can close that other thread. WP All Import is working fantastically, but when I try to map all those individual fields to the one "address" field for the pods-maps plugin/add-on, no data is received. Is there a simple solution that I'm missing?

jimtrue commented:
Sounds like it might be a 'structure' question (ie how to combine those fields into an array to be supported by the Map Field. @JoryHogeveen might be able to give some insight on this one. I'll close the other ticket: https://wordpress.org/support/topic/importing-to-address-field/#post-11139966

JoryHogeveen commented:
@bwdgraphics 
Please see: http://www.wpallimport.com/documentation/custom-fields/serialized-fields/

Also, I think it's better to create a new ticket for topics like this. Let's keep this thread purely for development purposes.

JoryHogeveen commented:
@tiptronic Since I couldn't wait for this to be actually finished and integrated into Pods I created a plugin to work with (actually 2 plugins, the address field and the maps component).
Keep in mind that if you are not comfortable with PHP (coding in general) I'm not sure it will be suitable.
Repo: https://github.com/JoryHogeveen/pods-address-maps

tiptronic commented:
@JoryHogeveen I am not really a PHP Pro, but I will definitely have a look at it asap, as I can definitively use this here and there. Thanks for sharing!

oligoform commented:
we run into issues with this in pods 2.8. This helps a little 
https://github.com/JoryHogeveen/pods-address-maps/commit/1af43f9d56a8e96adfef14f9b41689c9d3554608  but the geolocation doesn't work anymore.
maybe we could use /tribe-common/src/Tribe/Utils/Coordinates_Provider.php for this in pods 2.8? But how? I cannot choose it from the "shipped" pods fields.

JoryHogeveen committed:
Merge remote-tracking branch 'upstream/main' into feature/#657_address_map_field

# Conflicts:

JoryHogeveen committed:
Refactor Address field to new Pods structure and start on DFV UI

JoryHogeveen committed:
Improve Address UI with labels

JoryHogeveen committed:
Support custom input extras (maps component)

JoryHogeveen committed:
Auto rebuild assets

JoryHogeveen committed:
latlng formatting

JoryHogeveen committed:
Prepare LatLng UI (not available yet)

JoryHogeveen committed:
Make latlng fields opt-in for all address input types

JoryHogeveen committed:
Auto rebuild assets
